File: engine_lua/renderer/material.lua
require("lua_extension")
require("cpp_class")

--- @class Material 材质
Material = class("Material", CppClass)

function Material:ctor()
    Material.super.ctor(self)
end

--- 实例化C++ Class
function Material:InitCppClass()
    self.cpp_class_instance_ = Cpp.Material()
end

--- 加载Material文件并解析
function Material:Load(material_path)
    self.cpp_class_instance_:Load(material_path)
end

function Material:SetUniformMatrix4fv(shader_property_name, pointer)
    self.cpp_class_instance_:SetUniformMatrix4fv(shader_property_name, pointer)
end

function Material:SetUniform1i(shader_property_name, value)
    self.cpp_class_instance_:SetUniform1i(shader_property_name, value)
end

function Material:shader()
    return self.cpp_class_instance_:shader()
end

File: engine/renderer/material.cpp
//
// Created by Nemo li on 2023/12/6.
//

#include <app/application.h>
#include "rapidxml/rapidxml.hpp"
#include "rapidxml/rapidxml_print.hpp"
#include "rapidxml/rapidxml_utils.hpp"
#include "material.h"
#include "shader.h"
#include "texture2d.h"

Material::Material() {

}

Material::~Material() {
}

void Material::Load(const std::string &material_path) {
    //解析xml
    rapidxml::file<> xml_file((Application::data_path() + material_path).c_str());
    rapidxml::xml_document<> document;
    document.parse<0>(xml_file.data());

    //根节点
    rapidxml::xml_node<> *material_node = document.first_node("material");
    if (material_node == nullptr) {
        return;
    }

    rapidxml::xml_attribute<> *material_shader_attribute = material_node->first_attribute("shader");
    if (material_shader_attribute == nullptr) {
        return;
    }

    shader_ = Shader::Find(material_shader_attribute->value());

    //解析Texture
    rapidxml::xml_node<> *material_texture_node = material_node->first_node("texture");
    while (material_texture_node != nullptr) {
        rapidxml::xml_attribute<> *texture_name_attribute = material_texture_node->first_attribute("name");
        if (texture_name_attribute == nullptr) {
            return;
        }

        rapidxml::xml_attribute<> *texture_image_attribute = material_texture_node->first_attribute("image");
        if (texture_image_attribute == nullptr) {
            return;
        }

        std::string image_path = texture_image_attribute->value();
        textures_.emplace_back(texture_name_attribute->value(),
                               image_path.empty() ? nullptr : Texture2D::LoadFromFile(image_path));

        material_texture_node = material_texture_node->next_sibling("texture");
    }
}

void Material::SetUniformMatrix4fv(const std::string &shader_property_name, float *pointer) {

}

void Material::SetUniform1i(const std::string &shader_property_name, int value) {

}
